benchmarks/distributed/ddp/compare_ddp.py

```diff
@@ -0,0 +1,150 @@
+
+def create_ddp_model(module, rank, pg, ddp_option, buffer_size):
+    """Helper to create DDPModel. """
+    if ddp_option == DDPOption.DDP_CPP_CORE:
+        ddp_model = DDP(module, device_ids=[rank], process_group=pg)
+        ddp_model._set_static_graph()
+        return ddp_model
+    elif ddp_option == DDPOption.LEGACY_DISTRIBUTED_DATA_PARALLEL:
+        return legacy_ddp.LegacyDistributedDataParallel(module, pg, buffer_size)
+    elif ddp_option == DDPOption.PYTHON_DDP_SYNC:
+        return python_ddp.PythonDDP(module, pg, False, buffer_size)
+    elif ddp_option == DDPOption.PYTHON_DDP_ASYNC:
+        return python_ddp.PythonDDP(module, pg, True, buffer_size)
+    else:
+        raise NotImplementedError
+
+def run_ddp(rank, world_size, epochs, ddp_option, buffer_size):
+    print(f'Invoked run_ddp rank {rank}')
+
+    # Setup
+    print("setting up ... ")
+    setup(rank, world_size)
+    torch.manual_seed(rank)
+    torch.cuda.manual_seed(rank)
+    device = torch.device('cuda:%d' % rank)
+    print('done')
+
+    # Create ResNet50 module and wrap in DDP module.
+    pg = dist.distributed_c10d._get_default_group()
+    model = models.resnet50().to(device)
+    ddp_model = create_ddp_model(model, rank, pg, ddp_option, buffer_size)
+    assert ddp_model is not None
+
+    loss_fn = nn.MSELoss()
+    optimizer = optim.SGD(ddp_model.parameters(), lr=0.001)
+
+    # Container to hold: event -> list of events in milliseconds
+    MODEL_FORWARD = "forward"
+    MODEL_BACKWARD = "backward"
+    metrics = {MODEL_FORWARD: [], MODEL_BACKWARD: []}
+
+    for epoch in range(epochs):
+        if epoch % 10 == 0:
+            print(f'Epoch {epoch}/{epochs} ...')
+
+        start = torch.cuda.Event(enable_timing=True)
+        end = torch.cuda.Event(enable_timing=True)
+
+        # TODO(bowangbj): Switch to real training set from ImageNet.
+        inputs = torch.rand([32, 3, 224, 224], device=device)
+        labels = torch.rand([32, 1000], device=device)
+
+        # Forward
+        start.record()
+        outputs = ddp_model(inputs)
+        end.record()
+        torch.cuda.synchronize()
+        metrics[MODEL_FORWARD].append(start.elapsed_time(end))
+
+        # Backward
+        start.record()
+        loss_fn(outputs, labels).backward()
+        # Reduce all grad, this is needed for non-DDP_CPP_CORE since the hook
+        # for all_reduce does not exist yet.
+        if ddp_option != DDPOption.DDP_CPP_CORE:
+            ddp_model.all_reduce_grads()
+        optimizer.step()
+        optimizer.zero_grad()
+        end.record()
+        torch.cuda.synchronize()
+        metrics[MODEL_BACKWARD].append(start.elapsed_time(end))
+
+    if rank == 0:
+        print(f'\nMetrics for GPU {rank} ddp_option {ddp_option}:')
+        for step, elapsed_milliseconds in metrics.items():
+            A = np.array(elapsed_milliseconds)
+            print(' {N} iterations, {event}, mean={mean} ms, median={median} ms, p90={p90} ms, p99={p99} ms'.format(
```

Comment:
Shall we exclude the 1st 20 iterations to get rid of CUDACachingAllocator warmup overhead?

Reply:
Great point. Done.